pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/client/template.go

```diff
@@ -20,73 +20,252 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/tinkerbell/tink/protos/template"
+	tinkv1alpha1 "github.com/tinkerbell/tink/api/v1alpha1"
+	"gopkg.in/yaml.v3"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Template client for Tinkerbell.
+type Task struct {
+	Name        string            `json:"name"`
+	WorkerAddr  string            `json:"worker" yaml:"worker"`
```

Comment:
This was left on purpose because of a Tinkerbell issue in the template object. In the API it's called worker however in the Tinkerbell internal parser it still called workeraddr. Tinkerbell has two different templates, the Kubernetes one and the internal one. This is due to historical reasons that tink server used to run gRPC not kube.

Reply:
but the tags are the same..

Also can we leave a comment that these types are copied (I presume) from https://github.com/tinkerbell/tink/blob/main/internal/deprecated/workflow/types.go ?

Also that URL makes me wonder if we're building this new implementation based on deprecated types?

Reply:
This is not something that we can control, again if we remove that, things work! FYI we are using the latest tink stack and still this is a problem. This is a mitigation for a bug in the tink-contorller:
`,"error":"parsing yaml data: yaml: unmarshal errors:\n  line 6: field workeraddr not found in type workflow.Task",` this is still in the latest releases.

Reply:
Okay, the mismatch between the Go field "WorkerAddr" and the tag names "worker" are from upstream, fine.

But I still do not understand the tags here.

Upstream uses only YAML tags: https://github.com/tinkerbell/tink/blob/main/internal/deprecated/workflow/types.go#L13-L32

We copied (?) the types `Action` and `Task` from Tinkerbell, changed the YAML tags to JSON, and then added the `Actions` as YAML again to the `Template` struct. And ultimately we render templates as YAML.

So why the JSON fields for some structs?

Reply:
Not really! In the upstream API we also use JSON not only YAML. For example the hardware and template objects are JSON. The workflow type in v1alpha1 is the only thing that uses yaml. And even in the new v1alpha2 we started using JSON as well. And as I said this was mainly for historical reasons of how Tinkerbell object were handled, again Kubernetes was introduced for some time now but we still have some code which still leveraging the previous backend.